k8s/pkg/registryserver/nse.go

```diff
@@ -89,19 +104,220 @@ func (rs *nseRegistryService) RegisterNSE(ctx context.Context, request *registry
+	logger := span.Logger()
+
+	logger.Infof("Forwarding Bulk Register NSE stream...")
+
+	nsrURL := os.Getenv(ProxyNsmdK8sAddressEnv)
+	if strings.TrimSpace(nsrURL) == "" {
+		nsrURL = ProxyNsmdK8sAddressDefaults
+	}
+
+	ctx, cancel := context.WithCancel(srv.Context())
+	defer cancel()
+
+	remoteRegistry := nsmd.NewServiceRegistryAt(nsrURL)
+	defer remoteRegistry.Stop()
+
+	for {
+		stream, err := requestBulkRegisterNSEStream(ctx, remoteRegistry, nsrURL)
+		if err != nil {
+			logger.Warnf("Cannot connect to Proxy NSMGR %s : %v", nsrURL, err)
+			<-time.After(ProxyRegistryReconnectInterval)
+			continue
+		}
+
+		for {
+			request, err := srv.Recv()
+			if err != nil {
+				err = errors.Wrapf(err, "error receiving BulkRegisterNSE request : %v", err)
+				return err
+			}
+
+			logger.Infof("Forward BulkRegisterNSE request: %v", request)
+			err = stream.Send(request)
+			if err != nil {
+				logger.Warnf("Error forwarding BulkRegisterNSE request to %s : %v", nsrURL, err)
+				break
+			}
+		}
+
+		<-time.After(ProxyRegistryReconnectInterval)
+	}
+}
+
+func requestBulkRegisterNSEStream(ctx context.Context, remoteRegistry serviceregistry.ServiceRegistry, nsrURL string) (registry.NetworkServiceRegistry_BulkRegisterNSEClient, error) {
+	nseRegistryClient, err := remoteRegistry.NseRegistryClient(ctx)
+	if err != nil {
+		err = errors.Wrapf(err, "error forwarding BulkRegisterNSE request to %s : %v", nsrURL, err)
+		return nil, err
+	}
+
+	stream, err := nseRegistryClient.BulkRegisterNSE(ctx)
+	if err != nil {
+		err = errors.Wrapf(err, "error forwarding BulkRegisterNSE request to %s : %v", nsrURL, err)
+		return nil, err
+	}
+
+	return stream, nil
+}
+
 func (rs *nseRegistryService) RemoveNSE(ctx context.Context, request *registry.RemoveNSERequest) (*empty.Empty, error) {
 	st := time.Now()
 
-	logrus.Infof("Received RemoveNSE(%v)", request)
+	span := spanhelper.FromContext(ctx, "ProxyNsmgr.BulkRegisterNSE")
+	defer span.Finish()
+	logger := span.Logger()
+
+	logger.Infof("Received RemoveNSE(%v)", request)
 
 	if err := rs.cache.DeleteNetworkServiceEndpoint(request.GetNetworkServiceEndpointName()); err != nil {
 		return nil, err
 	}
-	logrus.Infof("RemoveNSE done: time %v", time.Since(st))
+
+	go func() {
+		if forwardErr := rs.forwardRemoveNSE(ctx, request); forwardErr != nil {
+			logger.Errorf("Cannot forward Remove NSE: %v", forwardErr)
+		}
+	}()
+
+	logger.Infof("RemoveNSE done: time %v", time.Since(st))
 	return &empty.Empty{}, nil
 }
+
+func (rs *nseRegistryService) forwardRegisterNSE(ctx context.Context, request *registry.NSERegistration) error {
+	span := spanhelper.FromContext(ctx, "ProxyNsmgr.BulkRegisterNSE")
+	defer span.Finish()
+	logger := span.Logger()
+
+	logger.Infof("Forwarding Register NSE request (%v)", request)
+
+	nsrURL := os.Getenv(ProxyNsmdK8sAddressEnv)
+	if strings.TrimSpace(nsrURL) == "" {
+		nsrURL = ProxyNsmdK8sAddressDefaults
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), ForwardingTimeout)
```

Comment:
Please use spanhelper.CopySpan() with context.Background() if you would like to have new context, or use span.Context() if same is ok.